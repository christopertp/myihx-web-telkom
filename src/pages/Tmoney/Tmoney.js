import React, { createContext, useCallback, useContext, useEffect, useRef, useState } from 'react';
import moment from 'moment';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useParams } from 'react-router-dom';
import Button from '../../components/elements/Button';
import IconButton from '../../components/elements/IconButton';
import InputPin from '../../components/elements/InputPin';
import Modal from '../../components/elements/Modal';
import Spinner from '../../components/elements/Spinner';
import FormTmoneyActivate from '../../components/forms/TmoneyActivate';
import WhiteCardPlain from '../../components/layouts/WhiteCardPlain';
import { AppContext } from '../../contexts';
import { thousand } from '../../utils/format';
import {
  clearHistory, fetchChangePin, fetchHistory, fetchLink,
  fetchReset, fetchSignup, pinFailedAction, clearFailedDialog
} from './actions';
import styles from './styles.scoped.css';

const Context = createContext({});

export default function Tmoney() {
  const [reset, setReset] = useState(false);
  const { page } = useParams();

  const contextValue = {
    reset, setReset
  };

  return (
    <Context.Provider value={contextValue}>
      {(p => {
        switch (p) {
          case 'activate': return <Activate />;
          case 'change-pin': return <ChangePin />;
          case 'link': return <Link />;
          default: return <Home />;
        }
      })(page)}
      <Overlay />
      <ErrorModal />
      <Reset />
    </Context.Provider>
  );
}

export function Activate() {
  const dispatch = useDispatch();
  const subtitle = 'Create a new password to activate wallet';
  const title = 'Activate Wallet';

  const onSubmit = (v) => {
    dispatch(fetchSignup(v.password));
  };

  return (
    <WhiteCardPlain icon="wallet" size="medium" subtitle={subtitle} title={title}>
      <FormTmoneyActivate onSubmit={onSubmit} />
    </WhiteCardPlain>
  );
}

export function ChangePin() {
  const [attempt, setAttempt] = useState(3);
  const [newPin, setNewPin] = useState('');
  const [pin, setPin] = useState('');
  const [step, setStep] = useState(0);

  const { setReset } = useContext(Context);
  const dispatch = useDispatch();
  const { isLoading, pinMessage } = useSelector(s => s.tmoney);

  const onConfirm = confirmPin => {
    if (newPin !== confirmPin) {
      setStep(1);
      dispatch(pinFailedAction('New PIN and confirmation PIN do not match'));
      return;
    }
    dispatch(fetchChangePin(pin, newPin, confirmPin));
  };

  const onEnterNewPin = value => {
    setNewPin(value);
    setStep(2);
    dispatch(pinFailedAction(''));
  };

  const onEnterPin = value => {
    setPin(value);
    setStep(1);
    dispatch(pinFailedAction(''));
  };

  const details = [
    {
      icon: 'verify',
      title: 'Enter Wallet PIN',
      subtitle: 'Enter your previous Wallet PIN to change your PIN',
      onSubmit: onEnterPin,
    },
    {
      icon: 'verify',
      title: 'Create a new PIN',
      subtitle: 'Create a new PIN for your Wallet',
      onSubmit: onEnterNewPin,
    },
    {
      icon: 'password',
      title: 'Confirm your new PIN',
      subtitle: 'Confirm your new PIN',
      onSubmit: onConfirm,
    },
  ];

  useEffect(() => {
    if (pinMessage === 'Wrong PIN') {
      setStep(0);
      setAttempt(attempt - 1);
    }
  }, [pinMessage]);

  const { icon, title, subtitle, onSubmit } = details[step];
  const errorMessage = pinMessage === 'Wrong PIN' ?
    `Wrong PIN! You have ${attempt} attempt(s) left` : pinMessage;

  return (
    <WhiteCardPlain back="/tmoney" icon={icon} size="medium" subtitle={subtitle} title={title}>
      <div className={styles.change}>
        <InputPin disabled={isLoading.s} error={errorMessage} onSubmit={onSubmit} />
        <Button disabled={isLoading.s} onClick={() => setReset(true)} variant="text">
          {isLoading.s ? <Spinner color="#ee3124" /> : 'I Forgot My PIN'}
        </Button>
      </div>
    </WhiteCardPlain>
  );
}

export function ErrorModal() {
  const dispatch = useDispatch();
  const history = useHistory();
  const { failedDialog } = useSelector(s => s.tmoney);
  const back = () => {
    dispatch(clearFailedDialog());
    history.push(failedDialog.to, failedDialog.notif);
  };
  return (
    <Modal className={styles.warning} onClose={back} open={!!failedDialog}>
      <img alt="warning" src="/assets/grfx_warning.svg" />
      <h1>{failedDialog.title}</h1>
      <p>{failedDialog.message}</p>
      <Button onClick={back}>OK</Button>
    </Modal>
  );
}

export function History() {
  const { histories } = useSelector(s => s.tmoney);
  return histories.map((v, i) => (
    <div className={styles.list} key={i}>
      <small>{moment(v.timestamp).format('DD MMMM YYYY hh:mm')}</small>
      <p>{v.transactionName}</p>
      <h3 className={v.amount < 0 ? styles.negative : styles.positive}>
        Rp{thousand(Math.abs(v.amount))}
      </h3>
    </div>
  ));
}

export function Home() {
  const [page, setPage] = useState(1);
  const { setReset } = useContext(Context);
  const dispatch = useDispatch();
  const history = useHistory();
  const prevY = useRef(0);
  const { balance, histories, historiesCompleted, isLoading } = useSelector(s => s.tmoney);

  const loadHistories = useCallback(page => {
    if (!historiesCompleted && !isLoading.d) {
      dispatch(fetchHistory(page));
      setPage(page + 1);
    }
  }, [historiesCompleted, isLoading.d, page]);

  useEffect(() => {
    loadHistories(page);
    return () => {
      dispatch(clearHistory());
    };
  }, []);

  useEffect(() => {
    const observer = new IntersectionObserver(([entry]) => {
      const { y } = entry.boundingClientRect;
      if (prevY.current > y) {
        loadHistories(page);
      }
      prevY.current = y;
    }, { threshold: 0.2 });
    const target = document.getElementsByTagName('footer')[0];
    observer.observe(target);
    return () => {
      observer.unobserve(target);
    };
  }, [loadHistories]);

  return (
    <WhiteCardPlain back="/profile" size="medium">
      <div className={styles.home}>
        <h3>Saldo</h3>
        {isLoading.d && !balance ? <span className="loading" /> : <h2>Rp{thousand(balance)}</h2>}
        <section className={styles.buttons}>
          <IconButton name="saldo_reset" onClick={() => setReset(true)} />
          <span>Reset PIN</span>
          <IconButton name="saldo_change" onClick={() => history.push('/tmoney/change-pin')} />
          <span>Change PIN</span>
          <IconButton name="saldo_trf" onClick={() => history.push('/linkaja')} />
          <span>Transfer to LinkAja</span>
        </section>
        <h3>Balance History</h3>
        {histories.length ? <History /> : <NoHistory />}
        {isLoading.d ?
          <div className={styles.loading}>
            <span className="loading" />
            <span className="loading" />
            <span className="loading" />
          </div> : null
        }
      </div>
    </WhiteCardPlain>
  );
}

export function Link() {
  const { setReset } = useContext(Context);
  const dispatch = useDispatch();
  const { isLoading, pinMessage } = useSelector(s => s.tmoney);

  const subtitle = 'This will link your T-Money Account to Indihome wallet. Enter your T-Money PIN to authorize this transaction.';

  const onSubmit = (pin) => {
    dispatch(fetchLink(pin));
  };

  return (
    <WhiteCardPlain back="/profile" icon="verify" size="medium" subtitle={subtitle} title="Enter Wallet PIN">
      <div className={styles.change}>
        <InputPin disabled={isLoading.s} error={pinMessage} onSubmit={onSubmit} />
        <Button disabled={isLoading.s} onClick={() => setReset(true)} variant="text">
          {isLoading.s ? <Spinner color="#ee3124" /> : 'I Forgot My PIN'}
        </Button>
      </div>
    </WhiteCardPlain>
  );
}

export function NoHistory() {
  const { isLoading } = useSelector(s => s.tmoney);
  if (isLoading.d) {
    return null;
  }
  return (
    <>
      <img alt="blank_default" src="/assets/grfx_blank_default.svg" />
      <p>Your balance history will show here when you<br />have a Saldo Transaction</p>
    </>
  );
}

export function Overlay() {
  const { setOverlay } = useContext(AppContext);
  const { isLoading } = useSelector(s => s.tmoney);

  useEffect(() => {
    setOverlay(isLoading.o);
  }, [isLoading.o]);

  return null;
}

export function Reset() {
  const { reset, setReset } = useContext(Context);
  const dispatch = useDispatch();
  const { isLoading, resetFinished } = useSelector(s => s.tmoney);

  const onReset = () => {
    dispatch(fetchReset());
  };

  useEffect(() => {
    resetFinished && setReset(false);
  }, [resetFinished]);

  return (
    <Modal className={styles.dialog} onClose={() => setReset(false)} open={reset}>
      <h1>Reset PIN?</h1>
      <p>
        Are you sure you want to reset your PIN?
        Your new PIN will be sent to your email and mobile number
      </p>
      <div>
        <Button disabled={isLoading.s} onClick={() => setReset(false)} variant="bordered">No</Button>
        <Button disabled={isLoading.s} onClick={() => onReset()}>
          {isLoading.s ? <Spinner /> : 'Yes'}
        </Button>
      </div>
    </Modal>
  );
}
