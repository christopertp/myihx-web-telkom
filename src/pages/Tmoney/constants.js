export const CLEAR_DIALOG = 'Tmoney/clear-dialog';
export const FAILED = 'Tmoney/failed';
export const HISTORY_CLEAR = 'Tmoney/history-clear';
export const HISTORY_FETCHED = 'Tmoney/history-fetched';
export const LOADING = 'Tmoney/loading';
export const PIN_FAILED = 'Tmoney/pin-failed';
export const RESET_FINISHED = 'Tmoney/reset-finished';
export const TRANSACTION_FAILED = 'Tmoney/transaction-failed';
