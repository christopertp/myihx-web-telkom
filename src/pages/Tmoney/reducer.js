import { CLEAR_DIALOG, FAILED, HISTORY_CLEAR, HISTORY_FETCHED, LOADING,
  PIN_FAILED, RESET_FINISHED, TRANSACTION_FAILED } from './constants';

const initialState = {
  balance: '',
  failedDialog: '',
  histories: [],
  historiesCompleted: false,
  isLoading: { d: false, o: false, s: false },
  message: '',
  pinMessage: '',
  resetFinished: false,
};

export default function reducer(state = initialState, action = {}) {
  const { type, balance, failedDialog, histories,
    isLoading, key, message, pinMessage, resetFinished } = action;

  switch (type) {
    case CLEAR_DIALOG:
      return {
        ...state,
        failedDialog: '',
      };
    case FAILED:
      return {
        ...state,
        message,
      };
    case HISTORY_CLEAR:
      return {
        ...state,
        histories: [],
        historiesCompleted: false,
      };
    case HISTORY_FETCHED:
      return {
        ...state,
        balance,
        histories: [ ...state.histories, ...histories ],
        historiesCompleted: histories.length < 10,
        isLoading: { ...state.isLoading, d: false },
      };
    case LOADING:
      return {
        ...state,
        isLoading: { ...state.isLoading, [key]: isLoading },
      };
    case PIN_FAILED:
      return {
        ...state,
        isLoading: { ...state.isLoading, s: false },
        pinMessage,
      };
    case RESET_FINISHED:
      return {
        ...state,
        isLoading: { ...state.isLoading, o: false, s: false },
        resetFinished,
      };
    case TRANSACTION_FAILED:
      return {
        ...state,
        failedDialog,
        isLoading: { ...state.isLoading, s: false },
      };
    default:
      return state;
  }
}
