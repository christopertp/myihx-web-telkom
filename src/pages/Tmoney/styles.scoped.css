.buttons {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  justify-items: center;
  margin: 1.5rem 0;
}

.buttons > button {
  background: linear-gradient(130deg, #470c08 0%, #b21f15 67%, #be3c1c 100%);
  border-radius: 1.5rem;
  grid-row: 1;
  height: 3.0625rem;
  padding: 0.75rem;
  width: 3.0625rem;
}

.buttons > span {
  color: #191a1a;
  font-size: 0.8125rem;
  grid-row: 2;
  line-height: 1.0625rem;
  margin-top: 0.3125rem;
  text-align: center;
}

.change > button {
  margin: 0 auto;
}

.dialog {
  text-align: center;
}

.dialog > button {
  margin-top: 1.3125rem;
  width: 100%;
}

.dialog > div {
  display: grid;
  grid-template-columns: 1fr 1fr;
  margin-top: 2.5625rem;
}

.dialog > div > button:first-child {
  margin-right: 0.4375rem;
}

.dialog > div > button:last-child {
  margin-left: 0.4375rem;
}

.dialog > h1 {
  color: #191a1a;
  font-size: 1.25rem;
  font-weight: bold;
  line-height: 1.625rem;
  margin-bottom: 0.5rem;
}

.dialog > p {
  color: #191a1a;
  font-size: 0.8125rem;
  line-height: 1.1875rem;
}

.home > h2 {
  color: #333333;
  font-size: 1.5rem;
  line-height: 1.875rem;
}

.home > h3 {
  color: #66686a;
  font-size: 0.75rem;
  font-weight: normal;
  letter-spacing: 0.0625rem;
  text-transform: uppercase;
}

.home > img {
  display: block;
  height: 10rem;
  margin: 3.4375rem auto 0;
}

.home > p {
  color: #191a1a;
  font-size: 0.8125rem;
  line-height: 1.1875rem;
  margin: 1.3125rem 0 3.4375rem;
  text-align: center;
}

.home > span {
  display: block;
  height: 1.8rem;
  width: 8rem;
}

.list {
  border-bottom: 0.0625rem solid #e5e6e6;
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: auto auto;
  margin-top: 0.25rem;
  padding: 0.4375rem 0;
}

.list > h3 {
  align-self: center;
  color: #191a1a;
  font-size: 0.9375rem;
  grid-column: 2;
  grid-row: 1/3;
  line-height: 1.1875rem;
  justify-self: end;
}

.list > h3.negative::before {
  content: '-';
}

.list > h3.positive {
  color: #0eb37e;
}

.list > h3.positive::before {
  content: '+';
}

.list > p {
  color: #333333;
  font-size: 0.9375rem;
  grid-row: 2;
  line-height: 1.4375rem;
}

.list > small {
  color: #66686a;
  font-size: 0.75rem;
  line-height: 1.125rem;
}

.loading > span {
  display: block;
  height: 2.5rem;
  margin: 0.75rem 0 0.5rem 0;
}

.warning {
  text-align: center;
}

.warning > button {
  width: 100%;
}

.warning > h1 {
  color: #191a1a;
  font-size: 1.25rem;
  line-height: 1.625rem;
  margin-bottom: 0.5rem;
}

.warning > img {
  margin-bottom: 1rem;
}

.warning > p {
  color: #191a1a;
  font-size: 0.8125rem;
  line-height: 1.1875rem;
  margin-bottom: 2.5rem;
}
