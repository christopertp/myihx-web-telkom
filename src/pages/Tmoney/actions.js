import { push } from 'react-router-redux';
import { tmoneyChangePin, tmoneyHistory, tmoneyLink, tmoneyResetPin, tmoneySignup } from '../../services/payment';
import { CLEAR_DIALOG, FAILED, HISTORY_CLEAR, HISTORY_FETCHED, LOADING, RESET_FINISHED, PIN_FAILED, TRANSACTION_FAILED } from './constants';

const transactionFailedDialog = {
  title: 'Transaction Failed',
  message: 'Please try again later',
};

export function fetchChangePin(pin, newPin) {
  return async dispatch => {
    dispatch(loadingAction('s', true));
    try {
      const { data } = await tmoneyChangePin(pin, newPin);
      if (data.status === 'SUCCESS') {
        dispatch(pinFailedAction(''));
        dispatch(push('/tmoney', { notif: 'PIN successfully changed!' }));
      } else {
        dispatch(pinFailedAction('Wrong PIN'));
        dispatch(push('/tmoney/change-pin'));
      }
    } catch (err) {
      const { data, message } = err;
      if (data.status === 'RESET_PIN') {
        const notif = [
          'Your PIN has been automatically reset.',
          'Your new PIN has been sent to your phone number',
        ].join(' ');
        dispatch(transactionFailedAction('Change PIN Failed', notif, '/tmoney', { notif }));
      } else {
        dispatch(pinFailedAction(message));
      }
    }
  };
}

export function fetchHistory(page) {
  return async dispatch => {
    dispatch(loadingAction('d', true));
    try {
      const { data } = await tmoneyHistory(page);
      const { balance, histories } = data;
      dispatch(historyFetchedAction(balance, histories));
    } catch (err) {
      dispatch(loadingAction('d', false));
      dispatch(failedAction(err.message));
    }
  };
}

export function fetchReset() {
  return async dispatch => {
    dispatch(resetFinishedAction(false));
    dispatch(loadingAction('o', true));
    dispatch(loadingAction('s', true));
    try {
      const { data } = await tmoneyResetPin();
      const { status } = data;
      dispatch(resetFinishedAction(true));
      if (status === 'SUCCESS') {
        const notif = 'New PIN has been successfully sent to your email and phone number!';
        dispatch(push('tmoney', { notif }));
      } else {
        const { title, message } = transactionFailedDialog;
        dispatch(transactionFailedAction(title, message, '/tmoney'));
      }
    } catch (err) {
      const { title, message } = transactionFailedDialog;
      dispatch(resetFinishedAction(true));
      dispatch(transactionFailedAction(title, message, '/tmoney'));
    }
  };
}

export function fetchLink(pin) {
  return async dispatch => {
    dispatch(loadingAction('s', true));
    try {
      const { data } = await tmoneyLink(pin);
      if (data.status === 'SUCCESS') {
        dispatch(push('/tmoney'));
      } else {
        const { title, message } = transactionFailedDialog;
        dispatch(transactionFailedAction(title, message, '/profile'));
      }
    } catch (err) {
      const { data } = err;
      const { title, message } = transactionFailedDialog;
      const blockedMessage = [
        'Your account has been blocked for your safety.',
        'Please contact us to unblock your account.',
        'Call +62 21 3808888 or email customer service@tmoney.co.id.',
      ].join(' ');

      switch (data.status) {
        case 'ACCOUNT_BLOCKED':
          dispatch(transactionFailedAction('Account Blocked', blockedMessage, '/profile'));
          break;
        case 'WRONG_PIN':
          dispatch(pinFailedAction('Wrong PIN'));
          break;
        default:
          dispatch(transactionFailedAction(title, message, '/profile'));
          break;
      }
    }
  };
}

export function fetchSignup(password) {
  return async dispatch => {
    dispatch(loadingAction('s', true));
    try {
      const { data } = await tmoneySignup(password);
      dispatch(loadingAction('s', false));
      dispatch(failedAction(''));
      if (data.status === 'SUCCESS') {
        dispatch(push('/tmoney', { notif: 'Wallet successfully activated!' }));
      } else {
        const { title, message } = transactionFailedDialog;
        dispatch(transactionFailedAction(title, message, '/profile'));
      }
    } catch (err) {
      dispatch(loadingAction('s', false));
      dispatch(failedAction(err.message));
    }
  };
}

export function clearFailedDialog() {
  return { type: CLEAR_DIALOG };
}

export function clearHistory() {
  return { type: HISTORY_CLEAR };
}

export function failedAction(message) {
  return { type: FAILED, message };
}

function historyFetchedAction(balance, histories) {
  return { type: HISTORY_FETCHED, balance, histories };
}

function loadingAction(key, isLoading) {
  return { type: LOADING, key, isLoading };
}

export function pinFailedAction(pinMessage) {
  return { type: PIN_FAILED, pinMessage };
}

function resetFinishedAction(resetFinished) {
  return { type: RESET_FINISHED, resetFinished };
}

export function transactionFailedAction(title, message, to, notif) {
  return { type: TRANSACTION_FAILED, failedDialog: { title, message, to, notif } };
}
