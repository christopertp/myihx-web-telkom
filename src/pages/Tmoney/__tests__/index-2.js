import React, { useCallback, useContext, useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import ShallowRenderer from 'react-test-renderer/shallow';
import Spinner from '../../../components/elements/Spinner';
import { Home, Link, NoHistory, Reset } from '../Tmoney';

const histories = [{
  timestamp: '2020-06-25 18:53:34',
  transactionName: 'Terima Transfer',
  amount: 15000,
}, {
  timestamp: '2020-06-25 16:36:03',
  transactionName: 'Isi Saldo TCASH',
  amount: -1000,
}];

useSelector.mockImplementation(fn => {
  fn({});
  return { balance: 10000000, histories, isLoading: { d: false, s: false } };
});

jest.mock('../actions', () => ({
  clearFailedDialog: jest.fn(),
  clearHistory: jest.fn(),
  fetchChangePin: jest.fn(),
  fetchLink: jest.fn(),
  fetchHistory: jest.fn(),
  fetchReset: jest.fn(),
  fetchSignup: jest.fn(),
  pinFailedAction: jest.fn(),
}));

describe('src/pages/Tmoney-2', () => {
  test('Home', () => {
    window.IntersectionObserver = jest.fn().mockImplementationOnce(() => ({
      observe: jest.fn(),
      unobserve: jest.fn(),
    }));

    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Home />);
    expect(tree).toMatchSnapshot();
  });

  test('Link', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Link />);
    expect(tree).toMatchSnapshot();
  });

  test('NoHistory', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<NoHistory />);
    expect(tree).toMatchSnapshot();
  });

  test('Reset', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Reset />);
    expect(tree).toMatchSnapshot();
  });

  test('Home buttons click', () => {
    const observe = jest.fn();
    const push = jest.fn();
    const setReset = jest.fn();
    const unobserve = jest.fn();

    useContext.mockImplementationOnce(() => ({ setReset }));
    useHistory.mockImplementationOnce(() => ({ push }));
    window.IntersectionObserver.mockImplementationOnce(() => ({ observe, unobserve }));

    const result = Home();
    const { children } = result.props.children.props;
    children[2].props.children[0].props.onClick();
    expect(setReset).toHaveBeenCalledWith(true);

    children[2].props.children[2].props.onClick();
    expect(push).toHaveBeenCalledWith('/tmoney/change-pin');

    children[2].props.children[4].props.onClick();
    expect(push).toHaveBeenCalledWith('/linkaja');
  });

  test('Home loading', () => {
    const isLoading = { d: true, s: false };
    const observe = jest.fn();
    const unobserve = jest.fn();

    useCallback.mockImplementationOnce(fn => fn, []);
    useEffect.mockImplementationOnce(fn => fn()())
      .mockImplementationOnce(fn => fn()());
    useSelector.mockImplementationOnce(() => ({ histories: [], isLoading }));
    window.IntersectionObserver.mockImplementationOnce(() => ({ observe, unobserve }));

    const result = Home();
    const { children } = result.props.children.props;
    expect(children[1].props.className).toBe('loading');
    expect(children[5].props.children[0].props.className).toBe('loading');
  });

  test('Home scroll', () => {
    const dispatch = jest.fn();
    const entry = { boundingClientRect: { y: 10 } };
    const observe = jest.fn();
    const unobserve = jest.fn();

    useDispatch.mockImplementationOnce(() => dispatch);
    useRef.mockImplementationOnce(() => ({ current: 100 }));
    window.IntersectionObserver.mockImplementationOnce(fn => {
      fn([entry]);
      return { observe, unobserve };
    });

    Home();
    expect(dispatch).toHaveBeenCalledTimes(2);

    useDispatch.mockImplementationOnce(() => dispatch);
    useRef.mockImplementationOnce(() => ({ current: 0 }));
    window.IntersectionObserver.mockImplementationOnce(fn => {
      fn([entry]);
      return { observe, unobserve };
    });

    Home();
    expect(dispatch).toHaveBeenCalledTimes(3);
  });

  test('Link loading', () => {
    const isLoading = { s: true };
    useSelector.mockImplementationOnce(() => ({ isLoading }));

    const result = Link();
    expect(result.props.children.props.children[1].props.children.type).toBe(Spinner);
  });

  test('Link reset pin', () => {
    const setReset = jest.fn();
    useContext.mockImplementationOnce(() => ({ setReset }));

    const result = Link();
    result.props.children.props.children[1].props.onClick();
    expect(setReset).toHaveBeenCalledWith(true);
  });

  test('Link submit', () => {
    const dispatch = jest.fn();
    useDispatch.mockImplementationOnce(() => dispatch);

    const result = Link();
    result.props.children.props.children[0].props.onSubmit();
    expect(dispatch).toHaveBeenCalledTimes(1);
  });

  test('NoHistory loading', () => {
    const isLoading = { d: true };
    useSelector.mockImplementationOnce(() => ({ isLoading }));

    const result = NoHistory();
    expect(result).toBe(null);
  });

  test('Reset finished', () => {
    const setReset = jest.fn();

    useContext.mockImplementationOnce(() => ({ setReset }));
    useSelector.mockImplementationOnce(() => ({ isLoading: { s: false }, resetFinished: true }));

    Reset();
    expect(setReset).toHaveBeenCalled();
  });

  test('Reset loading', () => {
    const setReset = jest.fn();

    useContext.mockImplementationOnce(() => ({ reset: true, setReset }));
    useSelector.mockImplementationOnce(() => ({ isLoading: { s: true } }));

    const result = Reset();
    expect(result.props.children[2].props.children[1].props.children.type).toBe(Spinner);
  });

  test('Reset on close', () => {
    const setReset = jest.fn();

    useContext.mockImplementationOnce(() => ({ reset: true, setReset }));

    const result = Reset();
    result.props.children[2].props.children[0].props.onClick();
    expect(setReset).toHaveBeenCalledWith(false);

    result.props.onClose();
    expect(setReset).toHaveBeenCalledWith(false);
    expect(setReset).toHaveBeenCalledTimes(2);
  });

  test('Reset on reset', () => {
    const dispatch = jest.fn();
    const setReset = jest.fn();

    useContext.mockImplementationOnce(() => ({ setReset }));
    useDispatch.mockImplementationOnce(() => dispatch);

    const result = Reset();
    result.props.children[2].props.children[1].props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });
});
