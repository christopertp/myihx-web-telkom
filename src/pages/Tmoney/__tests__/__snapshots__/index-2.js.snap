// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/Tmoney-2 Home 1`] = `
<WhiteCardPlain
  back="/profile"
  close=""
  help={false}
  icon=""
  size="medium"
  subtitle=""
  title=""
>
  <div>
    <h3>
      Saldo
    </h3>
    <h2>
      Rp
      10.000.000
    </h2>
    <section>
      <IconButton
        className=""
        name="saldo_reset"
        onClick={[Function]}
        title=""
        to=""
        type="button"
      />
      <span>
        Reset PIN
      </span>
      <IconButton
        className=""
        name="saldo_change"
        onClick={[Function]}
        title=""
        to=""
        type="button"
      />
      <span>
        Change PIN
      </span>
      <IconButton
        className=""
        name="saldo_trf"
        onClick={[Function]}
        title=""
        to=""
        type="button"
      />
      <span>
        Transfer to LinkAja
      </span>
    </section>
    <h3>
      Balance History
    </h3>
    <History />
  </div>
</WhiteCardPlain>
`;

exports[`src/pages/Tmoney-2 Link 1`] = `
<WhiteCardPlain
  back="/profile"
  close=""
  help={false}
  icon="verify"
  size="medium"
  subtitle="This will link your T-Money Account to Indihome wallet. Enter your T-Money PIN to authorize this transaction."
  title="Enter Wallet PIN"
>
  <div>
    <InputPin
      className=""
      disabled={false}
      error=""
      length={6}
      onSubmit={[Function]}
    />
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="text"
    >
      I Forgot My PIN
    </Button>
  </div>
</WhiteCardPlain>
`;

exports[`src/pages/Tmoney-2 NoHistory 1`] = `
<React.Fragment>
  <img
    alt="blank_default"
    src="/assets/grfx_blank_default.svg"
  />
  <p>
    Your balance history will show here when you
    <br />
    have a Saldo Transaction
  </p>
</React.Fragment>
`;

exports[`src/pages/Tmoney-2 Reset 1`] = `
<Modal
  className=""
  onClose={[Function]}
  open={false}
>
  <h1>
    Reset PIN?
  </h1>
  <p>
    Are you sure you want to reset your PIN? Your new PIN will be sent to your email and mobile number
  </p>
  <div>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="bordered"
    >
      No
    </Button>
    <Button
      className=""
      disabled={false}
      onClick={[Function]}
      to=""
      type="button"
      variant="red"
    >
      Yes
    </Button>
  </div>
</Modal>
`;
